### ISSUE TITLE
Updater.sh is not updating user.js

### ISSUE THREAD

**Issue by Opening-Button-8988** [arkenfox/user.js#1587]
I am using Arch Linux.

When I execute the updater.sh file, it does nothing. When I run it in the terminal, it shows 

```
This script will update to the latest user.js file and append any custom configurations from user-overrides.js. Continue Y/N? 
y
Status: user.js has been backed up and replaced with the latest version!
Status: Override file appended: user-overrides.js
```

Updater.sh is in the same location as the user-overrides.js, user.js, updater.bat and everything else that was in the zip file.

I know the .sh file is doing nothing because changes I make in user-overrides.js do not reflect to user.js, which is the intended purpose, right? If there's something I'm missing, please let me know.

**Comment by icpantsparti2** [comment]
Hope I am not stating the obvious, but my first thought is have you placed the updater.sh etc in your firefox profile directory (find that location by opening `about:support` page in firefox).

#### also add an entry at the end of your user-overrides.js, something like:

```
user_pref("_user.js.parrot", "SUCCESS: user-overrides.js");
```

you can check `about:config` page in firefox for that pref name when you re-open firefox after the script has run.

#### When you run the updater script look out for any error messages

```
# change directory to your firefox profile (edit as required)
cd /home/*/.mozilla/firefox/*.default
./updater.sh
```

#### Other commands that might help you see what's happening:

\# list some files (with permissions/size/date shown) (edit as required):

```
find /home/*/.mozilla -iregex "^.*/\(prefs.js\|user.js\|user-overrides.js\|updater.sh\)$" -printf '%Y %M %u:%g %#6s %TY-%Tm-%Td %TH:%TM:%.2TS %p\n'
```

\# see if overrides are showing on the end of user.js:

```
tail user.js
```

**Comment by TheNightRider12** [comment]
I too am having this issue. I emptied out my user-overrides to just have 
`user_pref("_user.js.parrot", "SUCCESS: user-overrides.js");`

Ran prefsCleaner, then ran updater.sh. All with Firefox closed, all from the correct directory. When I open Firefox and go into `about:config`, there is nothing showing up for "_user.js.parrot". 

I am on Fedora Linux 37.

**Comment by Thorin-Oakenpants** [comment]
> nothing is showing

well it should, because the user.js is riddled with them. You should at the very least get to `SUCCESS: No no he's not dead, he's, he's restin'!`

or do you mean it's not showing `SUCCESS: user-overrides.js`?

**Comment by TheNightRider12** [comment]
It's not showing anything when I search for "_user.js.parrot". It shows as the options when I search: "Boolean" then "Number" then "String". But no text like the other preferences I search for. 

Here is what I see:
![image](https://user-images.githubusercontent.com/8153239/203132135-1142f45b-28c2-4aa1-b9d9-b8673c577eb9.png)

**Comment by icpantsparti2** [comment]
In the Firefox profile directory, check file owner/permissions with your file manager, or in terminal with `ls -ld . *.js *.sh`

If you ran the updater with sudo then the user.js will be root/rw and then would not be loaded by the user you run Firefox under.  If that then try running the updater again without sudo: `./updater.sh`

**Comment by Onfroygmx** [comment]
Updater works fine, more a loading file issue:
![screenarkenfox](https://user-images.githubusercontent.com/25054585/203140558-4541ad48-1c1e-401d-9fc6-97ad29512d0a.png)

Also backup file is in the backup folder:
![image](https://user-images.githubusercontent.com/25054585/203140973-11ed1ed2-3826-4944-b2c5-851e002b470e.png)

To close 1589

Onf.

**Comment by rusty-snake** [comment]
https://github.com/arkenfox/user.js/wiki/3.4-Apply-&-Update-&-Maintain

> Put all the relevant files into the root folder of **your profile**

**Comment by TheNightRider12** [comment]
@rusty-snake Yes. I have used user.js before. Everything (updater.sh and prefsCleaner.sh) are in the root directory. I got to that directory by doing into `about:profiles` finding the currently in use profile, then going into the Root Directory from there. Everything is in there as it should be and has always been.

**Comment by earthlng** [comment]
so that's 3 people on linux with FF107 having the same issue where FF doesn't load the user.js?

Can you guys test if it loads a very simple user.js and without using the updater/prefscleaner scripts?
something like this fe:
```js
// comment on 1st line

user_pref("_user.js.parrot", "SUCCESS: test test test");
```

Anyone else have a Linux with FF107 where the user.js is loaded correctly?

I suspect mozilla changed something in their linux release and the user.js isn't loaded correctly or at all.

**Comment by 2glops** [comment]
Loading of user.js without issue here.
Using the updater/prefscleaner scripts, overrides parrot successfull.
FF 107 Linux Debian (and Nightly)

**Comment by earthlng** [comment]
installed a Fedora VM today and FF107 is not even in the repo yet?!?
Anyways, I used the updater to install the user.js and it worked with no problems. The parrot pref (and everything else) showed up in about:config.
I have no idea what the problem is. I think you guys will have to figure it out for yourselves, sorry

**Comment by devingoodman** [comment]
FWIW, I just ran arkenfox-cleaner.js, updater and prefscleaner on two machines: one running Linux Mint and the other running Fedora. Multiple FF profiles on each. Actual machines, not VMs. I had no problems at all running anything.

**Comment by Opening-Button-8988** [comment]
Apologies for abandoning this thread. It would work only if I ran the file in the terminal (without sudo). Executing the file in the file manager did not work. Yes, the file is definitely in the right place. It could be file permissions, but if that's the case, why would it work in the terminal without sudo but not in the file manager? It could be something to do with the file manager. I'll revisit this when I have the time.

**Comment by icpantsparti2** [comment]
@Opening-Button-8988, if you execute updater.sh from clicking in a file manager it would not progress past the "Continue Y/N?" confirmation part.  However...

<details><summary>(click me) unofficial/unsupported example</summary>

<br>
You could make another script which calls updater.sh, but you would need to be careful to check that the updater was successful (as you do not see any output info when clicking on it in a file manager).
<br><br>

```
#!/usr/bin/env bash
# "updater-wrapper.sh" to run arkenfox user.js updater with optional arguments
# you will need to make this executable eg: sudo chmod +x updater-wrapper.sh
# note: for help message listing all "Optional Arguments" run: ./updater.sh -h
./updater.sh -sv
```

</details>

**Comment by Opening-Button-8988** [comment]
@icpantsparti2 So the official method is via the terminal? If so, I'll use execute it in the terminal in the future.

I'm not familiar with bash scripting. Just not that skilled at linux yet. Since the official method is via the terminal, there's not much use for me to use the script, but thanks for letting me know there's a way it can be done in the file manager.

Should I close this thread?

**Comment by Thorin-Oakenpants** [comment]
> Should I close this thread?

No. I want to wait for #1595 to be evaluated, and depending on what you penguins say, the wiki could do with some better linux instructions

**Comment by Opening-Button-8988** [comment]
> > Should I close this thread?
> 
> No. I want to wait for #1595 to be evaluated, and depending on what you penguins say, the wiki could do with some better linux instructions

After some testing, I have absolutely no idea what was working before, but it is no longer working anymore.

Previously, I would run the updater.sh in the terminal successfully, but it did not reflect any changes to the user.js. Yes, the user.js and user-overrides.js were in the same folder as updater.sh. I moved them to a separate folder in /Downloads/ just in case something weird was going on with Firefox. It was still not reflecting any changes. I played around with permissions, including running the updater.sh as sudo and as a regular user. Neither worked.

**Comment by atomGit** [comment]
@Opening-Button-8988 - i might suggest starting over because this isn't making sense

first of all, never run anything using sudo unless it's necessary

1. delete _all_ the arkenfox stuff you downloaded - if you have a user-overrides.js, rename it temporarily 
2. go [here](https://raw.githubusercontent.com/arkenfox/user.js/master/user.js) and save the user.js to your FF profile where prefs.js is located (Ctr+S to save the file)
3. go [here](https://raw.githubusercontent.com/arkenfox/user.js/master/updater.sh) and save the updater.sh file in the same directory
4. open a terminal and cd to your FF profile directory: 
  `cd "$HOME/.mozilla/firefox/<PROFILE_FOLDER_NAME>"`
5. make updater.sh executable:
  `chmod +x updater.sh`
7. run the updater: 
  `./updater.sh -c`
9. start FF > about:config > search for '_user'

**Comment by Opening-Button-8988** [comment]
@atomGit Hi, thanks for responding.

I did all that. I even went to a separate machine (Linux Mint). I downloaded the zip from github, extracted the contents, put the contents in a new profile (root directory). I created a user-overrides.js (as an empty document, thus it is definitely .js), and made changes to the user-overrides.js. I just put something like:

`user_pref("browser.startup.page", 1);`

The user.js has `0` specified as default. I ran the updater.sh by right-clicking on it, clicking on "Run in Konsole" (I also tried other terminals). I typed "y" and enter following the prompt. It gave me the results:

```
Status: user.js has been backed up and replaced with the latest version!
Status: Override file appended: user-overrides.js
```
Yes, all the files are in the correct place. Yes, updater.sh is executable. I'm aware I only need to run files as sudo if it's necessary, but seeing as the file I wanted to work didn't work, I figured running it as sudo might work, cause that works sometimes. But I also ran it as regular user first, which did not work.

**Comment by Opening-Button-8988** [comment]
Okay, I figured out what was going wrong. I sincerely apologize for my stupidity.

I was assuming that the changes made in user-overrides.js would change the default values in user.js. I didn't realize user-overrides.js was **appended** (a word I kept hearing, but it hadn't clicked) to user.js, at the end of the file. This, along with realizing that the appended changes correctly reflect to about:config settings.

Apologies again. I should have known this.

**Comment by atomGit** [comment]
glad you figured it out - this probably should have given us a clue for helping you but i, for one, missed it...

> The updater.sh does update user.js with the latest version online, but does not reflect changes made in user-overrides.js.

**Comment by TheNightRider12** [comment]
> 
This fixed the issue for me as well. The user.js file got loaded successfully after following the steps you outlined.

**Comment by TheNightRider12** [comment]
@atomGit That step by step fix guide fixed the issue for me. User.js loaded for me now.

**Comment by Opening-Button-8988** [comment]
Update: I'm having an issue changing a specific user pref which is stubborn to changing from the user.js.

The pref is `browser.search.widget.inNavBar.` I have it listed in user-overrides.js as;

`user_pref("browser.search.widget.inNavBar", true);`

It's been correctly reflected in user.js. However, no such preference exists in prefs.js. Yes, I had the profile offline when I made the changes, I ran updater.sh and prefscleaner.sh successfully. After doing so, I launched the profile, and the changes were not reflected in about:config nor prefs.js, even after relaunching several times. The pref just won't change for some reason, while other prefs do.

Any advice?

**Comment by atomGit** [comment]
@TheNightRider12 - when you originally downloaded the user.js, exactly how did you do it?

**Comment by TheNightRider12** [comment]
@atomGit When I originally did it, I downloaded updater.sh, put it into the root directory. Also downloaded prefsCleaner.sh, put it in the same directory (root folder in my profile). Then made a user-overrides.js and added my stuff. Then I ran ./updater.sh. I do not think I used sudo when I did it originally.

**Comment by atomGit** [comment]
right, but where, exactly, did you download it from - did you go [here](https://github.com/arkenfox/user.js), right-click user.js and save as..?

or did you go [here](https://github.com/arkenfox/user.js/blob/master/user.js) and copy it?

or did you go [here](https://raw.githubusercontent.com/arkenfox/user.js/master/user.js) and copy/save it?

or did you download the whole repo?

or...

**Comment by Thorin-Oakenpants** [comment]
> Update: I'm having an issue changing a specific user pref which is stubborn to changing from the user.js.

Not all prefs are guaranteed to apply cleanly. Some are runtime (i.e they are set based on other values) and with quantum or just after they moved a lot of pref reading/setting to later in the process in their faster startup patches. Some cause issues when you try and use lock or other commands. Also check out #1582 and the related bugzilla - my best guess is it's something similar to this, where by the time the pref is read, the fast startup has already painted/drawn itself and it gets reset

user.js is not a high priority - they prefer policies for these sorts of things - and `browser.search.widget.inNavBar` is not an arkenfox pref. If you want it fixed, file a bugzilla

**Comment by TheNightRider12** [comment]
@atomGit So I did not download user.js myself. I downloaded updater.sh and prefsCleaner.sh. Then made my own user-overrides. Then let the updater.sh script download user.js and do everything else. That was originally when I first installed it to my firefox profile a few months back.
